# Projeto 4 - Analise de folhas de videira
**Entrega**

O projeto deve ser entregue até a terça-feira da 5ª semana de aula, envie o link do GitHub ou HTML na atividade no Moodle.

O prazo máximo é quarta-feira da 5ª semana de aula, valendo 20% a menos da nota.

**Dados do Certificado**

Linguagem: Python

Tecnologias: Pandas, Numpy, Tensorflow, Keras, PIL, Matplotlib e Sklearn

Carga horária: 20 horas

**Grapevine Leaves Image Dataset - Kaggle**

Vamos trabalhar com imagens de folhas de uvas que são usadas para vinho.

**Sobre este projeto**

Os principais pontos que serão avaliados:
- Levantamento de hipóteses
- Manipulação de imagens
- Criar um modelo de classificação

**Preparação do ambiente**

Faça o download e carregue o dataset Grapevine Leaves Image Dataset.

In [ ]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import zipfile
from PIL import Image

from keras.datasets import cifar10
from tensorflow.keras.models import Model
from tensorflow.keras.applications import resnet50, VGG19

from sklearn.tree import DecisionTreeClassifier
from sklearn.naive_bayes import GaussianNB, MultinomialNB, BernoulliNB


In [ ]:
from tensorflow import keras
import tensorflow as tf
import keras.backend as K
from tensorflow.keras.preprocessing.image import ImageDataGenerator
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import InputLayer, BatchNormalization, Dropout, Flatten, Dense, Activation, MaxPool2D, Conv2D
from tensorflow.keras.layers import Conv2D, Dense, Flatten, Dropout, Activation
from tensorflow.keras.layers import BatchNormalization, Reshape, MaxPooling2D, GlobalAveragePooling2D
from tensorflow.keras.callbacks import EarlyStopping, ModelCheckpoint, ReduceLROnPlateau

from PIL import Image
from PIL import UnidentifiedImageError

from sklearn.preprocessing import OneHotEncoder

from sklearn.model_selection import train_test_split

from sklearn.model_selection import KFold, StratifiedKFold   
from sklearn.model_selection import cross_val_score, cross_val_predict
from sklearn.metrics import accuracy_score, r2_score

from sklearn.metrics import classification_report, confusion_matrix
import itertools

import warnings
warnings.filterwarnings('ignore')

#Exercícios#
##Questão 1) Implemente uma rotina de carregamento das imagens (2 pontos)

Os datasets das imagens vem de arquivos 'zip', então usaremos o código abaixo para carregar.

In [ ]:
imgzip_Ak = zipfile.ZipFile("/content/Ak.zip")
Ak = imgzip_Ak.infolist()
imgzip_Ala_Idris = zipfile.ZipFile("/content/Ala_Idris.zip")
Ala_Idris = imgzip_Ala_Idris.infolist()
imgzip_Buzgulu = zipfile.ZipFile("/content/Buzgulu.zip")
Buzgulu = imgzip_Buzgulu.infolist()
imgzip_Dimnit = zipfile.ZipFile("/content/Dimnit.zip")
Dimnit = imgzip_Dimnit.infolist()
imgzip_Nazli = zipfile.ZipFile("/content/Nazli.zip")
Nazli = imgzip_Nazli.infolist()

Essa será a legenda usada para os nossos rótulos, cada número faz referência a um tipo de folha de uva, confome lista abaixo:

Ak - 0

Ala_Idris - 1

Buzgulu - 2

Dimnit - 3

Nazli - 4

##Questão 2) Faça a leitura das imagens e responda (2 pontos):

**a) Qual a resolução das imagens deste dataset? (0,5 ponto)**

**b) Quantas imagens por classe? (1,0 ponto)**

**c) Este dataset já está estruturado em conjuntos treinamento/teste ou deve-se adotar alguma metodologia na modelagem? (0,5 ponto)**

In [ ]:
img_w, img_h = 224, 224
X = []
y = []

In [ ]:
for f in Ak:
    ifile = imgzip_Ak.open(f)
    img_b = Image.open(ifile).convert('RGB')
    img = img_b.resize((img_w,img_h))
    X.append(np.asarray(img))
    y.append(0)
print(img_b)
print(img)
    #display(img)

<PIL.Image.Image image mode=RGB size=511x511 at 0x7FB83A188A50>
<PIL.Image.Image image mode=RGB size=224x224 at 0x7FB8C01B9950>


In [ ]:
for f in Ala_Idris:
    ifile = imgzip_Ala_Idris.open(f)
    img_b = Image.open(ifile).convert('RGB')
    img = img_b.resize((img_w,img_h))
    X.append(np.asarray(img))
    y.append(1)
print(img_b)
print(img)

<PIL.Image.Image image mode=RGB size=511x511 at 0x7FB8CBE67CD0>
<PIL.Image.Image image mode=RGB size=224x224 at 0x7FB8CBEC3650>


In [ ]:
for f in Buzgulu:
    ifile = imgzip_Buzgulu.open(f)
    img_b = Image.open(ifile).convert('RGB')
    img = img_b.resize((img_w,img_h))
    X.append(np.asarray(img))
    y.append(2)
print(img_b)
print(img)

<PIL.Image.Image image mode=RGB size=511x511 at 0x7FB84E18FBD0>
<PIL.Image.Image image mode=RGB size=224x224 at 0x7FB83A0E7390>


In [ ]:
for f in Dimnit:
    ifile = imgzip_Dimnit.open(f)
    img_b = Image.open(ifile).convert('RGB')
    img = img_b.resize((img_w,img_h))
    X.append(np.asarray(img))
    y.append(3)
print(img_b)
print(img)

<PIL.Image.Image image mode=RGB size=511x511 at 0x7FB83A023E10>
<PIL.Image.Image image mode=RGB size=224x224 at 0x7FB83A1A1950>


In [ ]:
for f in Nazli:
    ifile = imgzip_Nazli.open(f)
    img_b = Image.open(ifile).convert('RGB')
    img = img_b.resize((img_w,img_h))
    X.append(np.asarray(img))
    y.append(4)
print(img_b)
print(img)

<PIL.Image.Image image mode=RGB size=511x511 at 0x7FB83A16C1D0>
<PIL.Image.Image image mode=RGB size=224x224 at 0x7FB84E1B7750>


In [ ]:
X = np.array(X)
X.shape

(500, 224, 224, 3)

In [ ]:
y = np.array(y)
y.shape

(500,)

**a) Qual a resolução das imagens deste dataset? (0,5 ponto)**

In [ ]:
print('A resolução original do dataset:',img_b.size)
print('A resolução que utilizaremos:',img.size)

A resolução original do dataset: (511, 511)
A resolução que utilizaremos: (224, 224)


**b) Quantas imagens por classe? (1,0 ponto)**


In [ ]:
import collections
collections.Counter(y)

Counter({0: 100, 1: 100, 2: 100, 3: 100, 4: 100})

cada tipo de folha tem 100 imagens, logo temos um dataset bem equilibrado.

**c) Este dataset já está estruturado em conjuntos treinamento/teste ou deve-se adotar alguma metodologia na modelagem? (0,5 ponto)**

não, deve ser adotada uma metodologia na modelagem, que será o cross validation.

##Questão 3) Faça a preparação do dataset para extração de características com a CNN VGG-19. 

**Qual o formato do input da rede?**

**Verifique se há necessidade de transformação da imagem e, se sim, implemente esta transformação (1 ponto)**

**Qual o formato do input da rede?**

In [ ]:
input_shape=(224, 224, 3)

**Verifique se há necessidade de transformação da imagem e, se sim, implemente esta transformação (1 ponto)**

Houve necessidade de transformação para ter uma melhor acurácia, foi feita a transformação de 511X511 para 224X224

##Questão 4) Implemente o processo de extração de características utilizando a VGG-19 e a transformação deste espaço de característica de forma adequada (2 pontos)

In [ ]:
print("Loading the VGG19-ImageNet model ...")
model = VGG19(include_top=True, weights='imagenet', input_shape=(224, 224, 3), classes=1000,classifier_activation="softmax",pooling='max') 
model = Model(inputs=model.input, outputs=model.get_layer(index=-2).output)
model.summary()

##Questão 5) Utilize o espaço de características:

**a) Se o conjunto não estiver particionado, escolha uma técnica e justifique sua decisão. Implemente esta rotina (1 ponto)**

**b) Utilize os dados com os classificadores Árvores de Decisão e Naive-Bayes. Obtenha a acurácia dos modelos e avalie os resultados (2 pontos).**

In [ ]:
numero_folds = 5 # número de grupo, teoricamente quanto maior melhor, porem gera muito gasto, tempo para rodar
kf = KFold (n_splits = numero_folds, shuffle = True,random_state=10)  #esse garante que todos os grupos terão a mesma quantidade de cada tipo

for train_index, test_index in kf.split(X):
 #print('TRAIN:', train_index, 'TEST:', test_index)
 X_train, X_test = X[train_index], X[test_index]
 y_train, y_test = y[train_index], y[test_index]

In [ ]:
prediction = np.array(model.predict(X_train))
#X_train = np.reshape(prediction, (prediction.shape[0], (prediction.shape[1]*prediction.shape[2]*prediction.shape[3])))
X_train = np.reshape(prediction, (prediction.shape[0], prediction.shape[1]))

prediction = np.array(model.predict(X_test))
#X_test = np.reshape(prediction, (prediction.shape[0], (prediction.shape[1]*prediction.shape[2]*prediction.shape[3])))
X_test = np.reshape(prediction, (prediction.shape[0], prediction.shape[1]))

print('\tFeatures training shape: ', X_train.shape)
print('\tFeatures testing shape: ', X_test.shape)

	Features training shape:  (400, 4096)
	Features testing shape:  (100, 4096)


**a) Se o conjunto não estiver particionado, escolha uma técnica e justifique sua decisão. Implemente esta rotina (1 ponto)**

O conjunto foi particionado utilizando a técnica de cross validation com o Kfolds igual a 5. Desta maneira o conjunto vai ser treinado e testado de 5 maneiras difententes, 5 agrupamentos. 

In [ ]:
# #aplicar PCA

# from sklearn.decomposition import PCA

# pca = PCA( n_components=5)
# pca.fit(X_train)
# print(pca.explained_variance_ratio_,pca.explained_variance_ratio_.sum())
# X_train = pca.fit_transform(X_train)
# X_test = pca.transform(X_test)

Deixamos o a aplicação do PCA de lado neste momentos, ela nos trouxe uma piora muito significativa, acredito que para dataset maiores seria muito útil, porém para o nosso trabalho conseguiremos trabalhar sem e ter um acurácia melhor.

**b) Utilize os dados com os classificadores Árvores de Decisão e Naive-Bayes. Obtenha a acurácia dos modelos e avalie os resultados (2 pontos).**

In [ ]:
print("Classification with Decision tree ...")
DecTree = DecisionTreeClassifier(max_depth=5,criterion='entropy') # criterion='entropy'
Dtree = DecTree.fit(X_train, np.ravel(y_train, order='C'))
#result = DecTree.predict(X_test)

# acc = accuracy_score(result, y_test)
# print("\tAccuracy Decision Tree: %0.4f" % acc)

scores = cross_val_score(Dtree, X_test, y_test, cv=numero_folds)
print('Cross-validated scores:', scores)
print('Cross-validated scores:', scores.mean(),'+-', scores.std())

Classification with Decision tree ...
Cross-validated scores: [0.55 0.4  0.55 0.45 0.5 ]
Cross-validated scores: 0.49000000000000005 +- 0.058309518948453015


In [ ]:
print("Classification with Naive Bayes ...")
NB = MultinomialNB(alpha=1.5)
#NB = BernoulliNB(alpha=0.1,binarize=1) # alpha=0.5  binarize=0.5
#NB = GaussianNB()
Nbayes = NB.fit(X_train,np.ravel(y_train, order='C'))
#result = NB.predict(X_test)

# acc = accuracy_score(result, y_test)
# print("\tAccuracy Naive Bayes: %0.4f" % acc)

scores = cross_val_score(Nbayes, X_test, y_test, cv=numero_folds)
print('Cross-validated scores:', scores)
print('Cross-validated scores:', scores.mean(),'+-', scores.std())

Classification with Naive Bayes ...
Cross-validated scores: [0.65 0.7  0.65 0.55 0.7 ]
Cross-validated scores: 0.65 +- 0.05477225575051658


Como podemos ver, os resultados para o Naive Bayes foram bem melhores. Foram testados varios parâmetros dos dois modelos de classificação para chegar ao melhor resultado.

Para a Arvore de decisão chegamos a uma acurácia de 0.49, acredito que para imagens não é um resultado ruim, porém como conseguimos um resultado superior com o naive bayes, usaremos este como paramêtro para futuras melhorias no modelo.

O melhor resultado, então, foi o naive bayes com 0.65 de acurácia, um resultado consideravelmente bom, mas sabemos que podemos melhorar em futuros projetos, neste nosso segmento sempre podemos aprimorar nossos resultados, com outros modelos, paramêtros ou mais dados.

Para concluir, tivemos neste projeto a oportunidade de trabalhar com redes neurais, aprender como implementar acredito que seja o principal objetivo. Agora com mais pesquisas e mais experiência vamos aprimorando nosso conhecimento.